### PR TITLE
Typechecker improvements

### PR DESCRIPTION
**Pull request by e1528532**
## Basics

Several improvements regarding the typechecker module. I've reorganized the code a bit into smaller modules. Furthermore type inference has been implemented according to the specification, before it was only able to check keys where all regexes are known but didn't allow type variables to appear in intermediate terms. Also basically now almost everything can be decided using type inference without having to use a lot of explicit type signatures like it was before. And a lot of obsolete stuff from the old typechecker idea was removed as well.

## Checklist

- [ ] I added unit tests (the tests are like before but the underlying implementation changed)
- [X] I ran all tests locally and everything went fine
- [X] affected documentation is fixed
- [X] I added code comments, logging, and assertions (see doc/CODING.md)
- [X] meta data is updated (e.g. README.md of plugins)

### CODE REVIEW

**Comment by e1528532** [comment]
@markus2330 ready for a review / merge

File: doc/news/_preparation_next_release.md [hunk 1]
```diff
@@ -188,6 +188,12 @@ Thanks to Daniel Bugl.
   commonly used specification keywords to be used with the [typechecker](https://www.libelektra.org/plugins/typechecker). Currently the
   keywords `check/range`, `check/enum` and `check/validation` are supported. *(Armin Wurzinger)*
 
+### Typechecker
+
+- The plugin [typechecker](https://www.libelektra.org/plugins/typechecker) has been improved under the hood. It now supports a more concise
+and efficient typechecking process including a greatly improved type inference scheme that allows generate specification files and thus
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: allows us

File: doc/news/_preparation_next_release.md [hunk 2]
```diff
@@ -188,6 +188,12 @@ Thanks to Daniel Bugl.
...
+generated errors to be easier to understand. *(Armin Wurzinger)*
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: Ülease give an example of such a better error message.

File: doc/news/_preparation_next_release.md [hunk 3]
```diff
@@ -188,6 +188,12 @@ Thanks to Daniel Bugl.
   commonly used specification keywords to be used with the [typechecker](https://www.libelektra.org/plugins/typechecker). Currently the
   keywords `check/range`, `check/enum` and `check/validation` are supported. *(Armin Wurzinger)*
 
+### Typechecker
+
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md]: Shortly describe what the plugin does, not everyone knows it.

File: src/libs/typesystem/specelektra/Elektra/Parse.hs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import Data.Maybe    (fromMaybe)
+import Control.Monad (mapM, join)
+
+import Elektra.Types
+import qualified FiniteAutomata as FA
+
+import GhcPlugins
+import TyCoRep (Type(..), TyLit(..))
+
+-- This also ensures that every regex expression is a valid regex otherwise it returns EmptyRegex
```
Comment (markus2330) [inline src/libs/typesystem/specelektra/Elektra/Parse.hs]: What does "This" refer to? Why "also"? What else is ensured?

File: src/plugins/typechecker/Elektra/Typechecker.hs.in [hunk 5]
```diff
@@ -119,12 +117,15 @@ keyInterpreter specFile = do
     isFun (Fun _) = True
     isFun _ = False
 
-errorString :: InterpreterError -> String
-errorString (WontCompile es) = intercalate "\n" (header : map unbox es)
+errorString :: String -> InterpreterError -> String
+errorString f (WontCompile es) = intercalate "\n" (header : filterErrors es)
   where
-    header = "ERROR: Won't compile:"
+    header = "Typechecking or compilation failed:"
```
Comment (markus2330) [inline src/plugins/typechecker/Elektra/Typechecker.hs.in]: Can you separate type-checking and compilation errors?

File: src/plugins/typechecker/README.md [hunk 6]
```diff
@@ -63,59 +63,76 @@ specification and then prelude.
 
 In case a new plugin has been written which can be described using regular expressions, a
 plugin developer may want to teach the type system about this plugin. In order to do that,
-each configuration specification may contain a section prefixed with the path `elektra/spec`
-that contains information about how the type system should treat a specification keyword by
-formalizing its semantics as a function's type signature. Let's take the example of the
-fallback keyword, which would be formalized in the ini format as follows:
+each configuration specification may contain a section describing the effects of keywords.
+There are three metakeys that are interpreted by the typechecker, `elektra/spec/order`,
```
Comment (markus2330) [inline src/plugins/typechecker/README.md]: Please explain these three.

File: src/plugins/typechecker/README.md [hunk 7]
```diff
@@ -195,8 +212,5 @@ cabal install 'base >=4.9 && <4.12' 'containers >=0.5 && <0.6' \
 
 ## Limitations
 
-- Rather experimental
-- Type checking only happens when getting or setting
-a key in a mounted specification
+- Type checking only happens when getting or setting a key in a mounted specification
```
Comment (markus2330) [inline src/plugins/typechecker/README.md]: Which automatically happens if cascading keys are used, isn't it? This should be clarified here.

Comment (e1528532) [reply src/plugins/typechecker/README.md]: clarified

File: src/plugins/typechecker/README.md (hunk 6 above)
Comment (e1528532) [reply src/plugins/typechecker/README.md]: This is done in the readme already

File: src/plugins/typechecker/Elektra/Typechecker.hs.in (hunk 5 above)
Comment (e1528532) [reply src/plugins/typechecker/Elektra/Typechecker.hs.in]: Usually typechecking errors mean unresolveable constraints indicated by "Could not deduce". 
Maybe not the most reliable method as i can only check the error message at this point and have no further information about it, but i guess it should be fine enough.

File: src/libs/typesystem/specelektra/Elektra/Parse.hs (hunk 4 above)
Comment (e1528532) [reply src/libs/typesystem/specelektra/Elektra/Parse.hs]: this is the function in general, but also is unnecessary it doesn't do more than that )
also i don't normalize here anymore only during the constraint resolution that is a relict from experimentation as well

File: doc/news/_preparation_next_release.md (hunk 3 above)
Comment (e1528532) [reply doc/news/_preparation_next_release.md]: done

File: doc/news/_preparation_next_release.md (hunk 2 above)
Comment (e1528532) [reply doc/news/_preparation_next_release.md]: in the readme now, explained how to interprete it there as well, good point

File: doc/news/_preparation_next_release.md (hunk 1 above)
Comment (e1528532) [reply doc/news/_preparation_next_release.md]: adjusted

**Comment by e1528532** [comment]
@markus2330 next try ;)

**Comment by markus2330** [comment]
Thank you, great job! Do you have now some time to fix pluginprocess? It also needs some more unit tests so that we can be sure that the forked plugin gets all parameters correctly.

**Pull request merged**